packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.base.tsx

```diff
@@ -337,53 +336,52 @@ export class DetailsHeaderBase extends BaseComponent<IDetailsHeaderBaseProps, ID
   }
 
   private _onDrop(item?: any, event?: DragEvent): void {
-    const draggedColumnIndex = this._draggedColumnIndex;
     const { columnReorderProps } = this.state;
+
     // Target index will not get changed if draggeditem is after target item.
     if (this._draggedColumnIndex >= 0 && event) {
-      const targetIndex = draggedColumnIndex > this._currentDropHintIndex! ? this._currentDropHintIndex! : this._currentDropHintIndex! - 1;
+      const targetIndex =
+        this._draggedColumnIndex > this._currentDropHintIndex! ? this._currentDropHintIndex! : this._currentDropHintIndex! - 1;
       let isValidDrop = false;
       event.stopPropagation();
       if (this._isValidCurrentDropHintIndex()) {
         isValidDrop = true;
-        this._onDropIndexInfo.sourceIndex = draggedColumnIndex;
+        this._onDropIndexInfo.sourceIndex = this._draggedColumnIndex;
         this._onDropIndexInfo.targetIndex = targetIndex;
       }
-      this._resetDropHints();
-      this._dropHintDetails = {};
-      this._draggedColumnIndex = -1;
       if (isValidDrop) {
         if (columnReorderProps && columnReorderProps.onColumnDrop) {
           const dragDropDetails: IColumnDragDropDetails = {
-            draggedIndex: draggedColumnIndex,
+            draggedIndex: this._draggedColumnIndex,
             targetIndex: targetIndex
           };
           columnReorderProps.onColumnDrop(dragDropDetails);
         } else if (columnReorderProps && columnReorderProps.handleColumnReorder) {
-          columnReorderProps.handleColumnReorder(draggedColumnIndex, targetIndex);
+          columnReorderProps.handleColumnReorder(this._draggedColumnIndex, targetIndex);
         }
       }
     }
+
+    this._resetDropHints();
+    this._dropHintDetails = {};
+    this._draggedColumnIndex = -1;

```

Comment:
This would cause a run-time error with an `undefined` column if set earlier now that `this._draggedColumnIndex` is the source of truth. I moved it to the end of the `_onDrop` function because we should always reset the drop "state" on completion of the drop.